[SOUP] Port SoupNetworkProxySettings to new IPC serialization

csaavedra opened the pull request:
#### 1f7e6f818b85efbef6b8f056e86e683cfc859779
<pre>
[SOUP] Port SoupNetworkProxySettings to new IPC serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=272085">https://bugs.webkit.org/show_bug.cgi?id=272085</a>

Reviewed by Michael Catanzaro.

Move the serialization of the SoupNetworkProxySettingsMode
to WebKit/Shared to a single file for soup-related argument
coders. Also add a serializer for GUniquePtr&lt;char*&gt;, as it&apos;s
needed for the ignored hosts list. With this in place,
add a generator for SoupNetworkProxySettings and remove
the handmade coder.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/network/soup/SoupNetworkProxySettings.h:
(WebCore::SoupNetworkProxySettings::SoupNetworkProxySettings):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/glib/ArgumentCodersGlib.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/soup/WebCoreArgumentCodersSoup.cpp: Removed.
* Source/WebKit/Shared/soup/WebCoreArgumentCodersSoup.serialization.in: Renamed from Source/WebCore/platform/network/soup/SoupNetworkProxySettings.serialization.in.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:

Canonical link: <a href="https://commits.webkit.org/277107@main">https://commits.webkit.org/277107@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/923c5b51966d1c1e81e44869fc51a254fa6b1748

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46618 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25775 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49225 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49295 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42663 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48925 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23239 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38020 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47197 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22761 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40171 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19280 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20149 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41288 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4663 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/42923 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41666 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51197 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21625 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18022 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45306 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22915 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44259 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/6539 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22618 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/network/soup/SoupNetworkProxySettings.h:
```diff
@@ -48,6 +48,14 @@ struct SoupNetworkProxySettings {
     {
     }
 
+    SoupNetworkProxySettings(Mode proxyMode,  const CString& defaultURL, const GUniquePtr<char*>& hosts, const HashMap<CString, CString>& map)
```
```suggestion
    SoupNetworkProxySettings(Mode proxyMode, const CString& defaultURL, const GUniquePtr<char*>& hosts, const HashMap<CString, CString>& map)
```

mcatanzaro commented on Source/WebKit/Platform/IPC/glib/ArgumentCodersGlib.h:
```diff
@@ -0,0 +1,59 @@
+
+namespace IPC {
+
+template<> struct ArgumentCoder<GUniquePtr<char*>> {
```
Include headers for what you use, please. E.g. glib.h, GUniquePtr.h, CString.h...

mcatanzaro commented on Source/WebKit/Platform/IPC/glib/ArgumentCodersGlib.h:
```diff
@@ -0,0 +1,59 @@
+        if (!length)
+            return std::nullopt;
+
+        auto* strv = g_new0(char*, *length + 1);
```
Oops, it's leaked if `!strOptional`. Better to always use the smart pointer for the entire scope:

```suggestion
        GUniquePtr<char*> strv(g_new0(char*, *length + 1));
```

csaavedra replied:
Good point!

csaavedra committed:
[SOUP] Port SoupNetworkProxySettings to new IPC serialization
https://bugs.webkit.org/show_bug.cgi?id=272085

Reviewed by Michael Catanzaro.

Move the serialization of the SoupNetworkProxySettingsMode
to WebKit/Shared to a single file for soup-related argument
coders. Also add a serializer for GUniquePtr<char*>, as it's
needed for the ignored hosts list. With this in place,
add a generator for SoupNetworkProxySettings and remove
the handmade coder.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/platform/network/soup/SoupNetworkProxySettings.h:
(WebCore::SoupNetworkProxySettings::SoupNetworkProxySettings):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
* Source/WebKit/Platform/IPC/glib/ArgumentCodersGlib.h: Added.
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/Shared/soup/WebCoreArgumentCodersSoup.cpp: Removed.
* Source/WebKit/Shared/soup/WebCoreArgumentCodersSoup.serialization.in: Renamed from Source/WebCore/platform/network/soup/SoupNetworkProxySettings.serialization.in.
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:

Canonical link: https://commits.webkit.org/277107@main
